src/binary_space_partition.rs

```diff
@@ -0,0 +1,461 @@
+        iter
+    }
+
+    // set the current node to the one provided
+    // and add any child leaves to the node vec
+    fn add_subtrees(&mut self, node: &'a Leaf) {
+        if let Some(ref left) = node.left_child {
+            self.right_nodes.push(&*left);
+        }
+        if let Some(ref right) = node.right_child {
+            self.right_nodes.push(&*right);
+        }
+
+        self.current_node = Some(node);
+    }
+}
+
+impl<'a> Iterator for LeafIterator<'a> {
+    type Item = &'a Leaf;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let result = self.current_node.take();
+        if let Some(rest) = self.right_nodes.pop() {
+            self.add_subtrees(rest);
+        }
+
+        match result {
+            Some(leaf) => Some(&*leaf),
+            None => None,
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+pub enum TileType {
+    Space = 0,
+    Floor = 1,
+    Wall = 2,
+}
+pub struct Level {
+    width: usize,
+    height: usize,
+    board: Vec<Vec<usize>>,
+    rooms: Vec<Room>,
+    increment: usize,
+}
+
+impl Level {
+    fn new(width: usize, height: usize) -> Self {
+        let mut new_level = Level {
+            width,
+            height,
+            board: Vec::new(),
+            rooms: Vec::new(),
+            increment: 0,
+        };
+        new_level.update_board();
+        new_level
+    }
+
+    fn update_board(&mut self) -> Vec<Vec<usize>> {
+        let mut new_board = Vec::new();
+        self.increment += 1;
+        for _ in 0..self.height {
+            let gen_floor_first = true;
+
+            let mut row = vec![TileType::Floor as usize; self.width as usize];
+            if !gen_floor_first {
+                row = vec![TileType::Space as usize; self.width as usize];
+            }
+
+            new_board.push(row);
+        }
+        for room in &self.rooms {
+            for row in 0..room.height {
+                for col in 0..room.width {
+                    let y = (room.y + row) as usize;
+                    let x = (room.x + col) as usize;
+                    if row == 0 || col == 0 || row == room.height - 1 || col == room.width - 1 {
+                        // might just let byond handle the walls
+                        new_board[y][x] = TileType::Wall as usize;
+                    } else {
+                        new_board[y][x] = TileType::Floor as usize;
+                    }
+                }
+            }
+        }
+        self.board = new_board.clone();
+        new_board
+    }
+
+    fn add_room(&mut self, room: &Room) {
+        self.rooms.push(room.clone());
+        self.update_board();
+    }
+}
+
+impl fmt::Display for Level {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        for row in 0..self.height as usize {
+            for col in 0..self.width as usize {
+                write!(f, "{}", self.board[row][col])?
+            }
+            write!(f, "\n")?
+        }
+
+        Ok(())
+    }
+}
```

Comment:
Is this required? I think Arrays in rust already implement the `Debug` trait so if you did need to print this out you don't necessarily need to do a custom Display trait implementation